lib/compiler/src/compiler.rs

```diff
@@ -63,47 +63,240 @@ where
+
+        // Create a parser with all features enabled to detect what the module actually uses
         let mut wasm_features = WasmFeatures::default();
-        wasm_features.set(WasmFeatures::BULK_MEMORY, features.bulk_memory);
-        wasm_features.set(WasmFeatures::THREADS, features.threads);
-        wasm_features.set(WasmFeatures::REFERENCE_TYPES, features.reference_types);
-        wasm_features.set(WasmFeatures::MULTI_VALUE, features.multi_value);
-        wasm_features.set(WasmFeatures::SIMD, features.simd);
-        wasm_features.set(WasmFeatures::TAIL_CALL, features.tail_call);
-        wasm_features.set(WasmFeatures::MULTI_MEMORY, features.multi_memory);
-        wasm_features.set(WasmFeatures::MEMORY64, features.memory64);
-        wasm_features.set(WasmFeatures::EXCEPTIONS, features.exceptions);
-        wasm_features.set(WasmFeatures::EXTENDED_CONST, features.extended_const);
-        wasm_features.set(WasmFeatures::RELAXED_SIMD, features.relaxed_simd);
+        // Enable all features for detection
+        wasm_features.set(WasmFeatures::BULK_MEMORY, true);
+        wasm_features.set(WasmFeatures::THREADS, true);
+        wasm_features.set(WasmFeatures::REFERENCE_TYPES, true);
+        wasm_features.set(WasmFeatures::MULTI_VALUE, true);
+        wasm_features.set(WasmFeatures::SIMD, true);
+        wasm_features.set(WasmFeatures::TAIL_CALL, true);
+        wasm_features.set(WasmFeatures::MULTI_MEMORY, true);
+        wasm_features.set(WasmFeatures::MEMORY64, true);
+        wasm_features.set(WasmFeatures::EXCEPTIONS, true);
+        wasm_features.set(WasmFeatures::EXTENDED_CONST, true);
+        wasm_features.set(WasmFeatures::RELAXED_SIMD, true);
         wasm_features.set(WasmFeatures::MUTABLE_GLOBAL, true);
         wasm_features.set(WasmFeatures::SATURATING_FLOAT_TO_INT, true);
         wasm_features.set(WasmFeatures::FLOATS, true);
         wasm_features.set(WasmFeatures::SIGN_EXTENSION, true);
-        wasm_features.set(WasmFeatures::GC_TYPES, true);
-
-        // Not supported
-        wasm_features.set(WasmFeatures::COMPONENT_MODEL, false);
-        wasm_features.set(WasmFeatures::FUNCTION_REFERENCES, false);
-        wasm_features.set(WasmFeatures::MEMORY_CONTROL, false);
-        wasm_features.set(WasmFeatures::GC, false);
-        wasm_features.set(WasmFeatures::COMPONENT_MODEL_VALUES, false);
-        wasm_features.set(WasmFeatures::COMPONENT_MODEL_NESTED_NAMES, false);
-
-        let mut validator = Validator::new_with_features(wasm_features);
-        validator
-            .validate_all(data)
-            .map_err(|e| CompileError::Validate(format!("{e}")))?;
+
+        // Parse the module and check for feature usage
+        for payload in Parser::new(0).parse_all(data) {
+            match payload {
+                Ok(Payload::TagSection(_)) => {
+                    // The presence of a tag section means exceptions are used
+                    detected_features.exceptions = true;
```

Comment:
Makes sense for now, but it should be noted that albeit being currently used for eh only, `tags` could be used for other features (or non-eh related endeavours) in the future, as noted [in the proposal](https://github.com/WebAssembly/exception-handling/blob/main/proposals/exception-handling/Exceptions.md#exception-handling-1).